### PR TITLE
Providermanager cleanup

### PR DESCRIPTION
**Pull request by 1337joe**
A bunch of incremental improvements to ProviderManager.cs and related files, primarily focusing on `RefreshSingleItem`, `GetImageProviders`, and `GetMetadataProviders`.

**Changes**
- Pulled a couple public methods up to the interface, resolved `#nullable disable`, general minor cleanup
- Stored library type options and fetcher order logic up front instead of rechecking multiple times for each provider
- Merged and simplified similar/duplicated code for image and metadata sorting
- Added unit tests

### CODE REVIEW

File: MediaBrowser.Providers/Manager/ProviderManager.cs [hunk 1]
```diff
@@ -361,165 +361,85 @@ public IEnumerable<IMetadataProvider<T>> GetMetadataProviders<T>(BaseItem item,
...
+                {
+                    // local and remote providers will be interleaved in the final order
+                    // only relative order within a type matters: consumers of the list filter to one or the other
+                    switch (i)
+                    {
+                        case ILocalMetadataProvider: return GetConfiguredOrder(localMetadataReaderOrder, i.Name);
+                        case IRemoteMetadataProvider: return GetConfiguredOrder(metadataFetcherOrder, i.Name);
+                        default: return int.MaxValue; // default to end
+                    }
+                })
```
Comment (crobibero) [inline MediaBrowser.Providers/Manager/ProviderManager.cs:380]:
```suggestion
                    // local and remote providers will be interleaved in the final order
                    // only relative order within a type matters: consumers of the list filter to one or the other
                    i switch
                    {
                        ILocalMetadataProvider => GetConfiguredOrder(localMetadataReaderOrder, i.Name),
                        IRemoteMetadataProvider => GetConfiguredOrder(metadataFetcherOrder, i.Name),
                        // Default to end
                        _ => int.MaxValue
                    })
```

**Comment by crobibero** [comment]
@jellyfin-bot rebase

**Review by crobibero — APPROVED** [review]

File: tests/Jellyfin.Providers.Tests/Manager/ProviderManagerTests.cs [hunk 2]
```diff
@@ -0,0 +1,620 @@
...
+            for (var i = 0; i < servicesList.Length; i++)
+            {
+                if (i == expectedIndex)
+                {
+                    servicesList[i].Verify(mock => mock.RefreshMetadata(It.IsAny<BaseItem>(), It.IsAny<MetadataRefreshOptions>(), It.IsAny<CancellationToken>()), Times.Once());
+                }
+                else
+                {
+                    servicesList[i].Verify(mock => mock.RefreshMetadata(It.IsAny<BaseItem>(), It.IsAny<MetadataRefreshOptions>(), It.IsAny<CancellationToken>()), Times.Never());
+                }
```
Comment (cvium) [inline tests/Jellyfin.Providers.Tests/Manager/ProviderManagerTests.cs:97]:
```suggestion
                var times = i == expectedIndex ? Times.Once() : Times.Never();
                servicesList[i].Verify(mock => mock.RefreshMetadata(It.IsAny<BaseItem>(), It.IsAny<MetadataRefreshOptions>(), It.IsAny<CancellationToken>()), times);
```

**Review by cvium — APPROVED** [review]

File: tests/Jellyfin.Providers.Tests/Manager/ProviderManagerTests.cs [hunk 3]
```diff
@@ -0,0 +1,614 @@
...
+// Allow Moq to see internal class
+[assembly: InternalsVisibleTo("DynamicProxyGenAssembly2")]
+
+namespace Jellyfin.Providers.Tests.Manager
+{
+    public class ProviderManagerTests
+    {
+        private static readonly ILogger<ProviderManager> _logger = new NullLogger<ProviderManager>();
+
+        private static TheoryData<Mock<IMetadataService>[], int> RefreshSingleItemOrderData()
```
Comment (Bond-009) [inline tests/Jellyfin.Providers.Tests/Manager/ProviderManagerTests.cs:33]:
```suggestion
        public static TheoryData<Mock<IMetadataService>[], int> RefreshSingleItemOrderData()
```

File: tests/Jellyfin.Providers.Tests/Manager/ProviderManagerTests.cs [hunk 4]
```diff
@@ -0,0 +1,614 @@
...
+            var servicesList = new[] { MockIMetadataService(false, serviceFound) };
+
+            using var providerManager = GetProviderManager();
+            AddParts(providerManager, metadataServices: servicesList.Select(s => s.Object).ToArray());
+
+            var refreshOptions = new MetadataRefreshOptions(Mock.Of<IDirectoryService>(MockBehavior.Strict));
+            var actual = providerManager.RefreshSingleItem(item, refreshOptions, CancellationToken.None);
+
+            var expectedResult = serviceFound ? ItemUpdateType.MetadataDownload : ItemUpdateType.None;
+            Assert.Equal(expectedResult, actual.Result);
```
Comment (Bond-009) [inline tests/Jellyfin.Providers.Tests/Manager/ProviderManagerTests.cs:111]:
```suggestion
        public async Task RefreshSingleItem_ServiceOrdering_FollowsPriority(Mock<IMetadataService>[] servicesList, int expectedIndex)
        {
            var item = new Movie();

            using var providerManager = GetProviderManager();
            AddParts(providerManager, metadataServices: servicesList.Select(s => s.Object).ToArray());

            var refreshOptions = new MetadataRefreshOptions(Mock.Of<IDirectoryService>(MockBehavior.Strict));
            var actual = await providerManager.RefreshSingleItem(item, refreshOptions, CancellationToken.None).ConfigureAwait(false);

            Assert.Equal(ItemUpdateType.MetadataDownload, actual);
            for (var i = 0; i < servicesList.Length; i++)
            {
                var times = i == expectedIndex ? Times.Once() : Times.Never();
                servicesList[i].Verify(mock => mock.RefreshMetadata(It.IsAny<BaseItem>(), It.IsAny<MetadataRefreshOptions>(), It.IsAny<CancellationToken>()), times);
            }
        }

        [Theory]
        [InlineData(true)]
        [InlineData(false)]
        public async Task RefreshSingleItem_RefreshMetadata_WhenServiceFound(bool serviceFound)
        {
            var item = new Movie();

            var servicesList = new[] { MockIMetadataService(false, serviceFound) };

            using var providerManager = GetProviderManager();
            AddParts(providerManager, metadataServices: servicesList.Select(s => s.Object).ToArray());

            var refreshOptions = new MetadataRefreshOptions(Mock.Of<IDirectoryService>(MockBehavior.Strict));
            var actual = await providerManager.RefreshSingleItem(item, refreshOptions, CancellationToken.None).ConfigureAwait(false);

            var expectedResult = serviceFound ? ItemUpdateType.MetadataDownload : ItemUpdateType.None;
            Assert.Equal(expectedResult, actual);
```

File: tests/Jellyfin.Providers.Tests/Manager/ProviderManagerTests.cs [hunk 5]
```diff
@@ -0,0 +1,614 @@
...
+        }
+
+        private static TheoryData<int, int[]?, int[]?, int?[]?, int[]> GetImageProvidersOrderData()
```
Comment (Bond-009) [inline tests/Jellyfin.Providers.Tests/Manager/ProviderManagerTests.cs:114]:
```suggestion
        public static TheoryData<int, int[]?, int[]?, int?[]?, int[]> GetImageProvidersOrderData()
```

File: tests/Jellyfin.Providers.Tests/Manager/ProviderManagerTests.cs [hunk 6]
```diff
@@ -0,0 +1,614 @@
...
+
+            using var providerManager = GetProviderManager(baseItemManager: baseItemManager.Object);
+            AddParts(providerManager, imageProviders: new[] { provider });
+
+            var actualProviders = providerManager.GetImageProviders(item, refreshOptions).ToArray();
+
+            Assert.Equal(expected ? 1 : 0, actualProviders.Length);
+        }
+
+        private static TheoryData<string[], int[]?, int[]?, int[]?, int[]?, int?[]?, int[]> GetMetadataProvidersOrderData()
```
Comment (Bond-009) [inline tests/Jellyfin.Providers.Tests/Manager/ProviderManagerTests.cs:239]:
```suggestion
        public static TheoryData<string[], int[]?, int[]?, int[]?, int[]?, int?[]?, int[]> GetMetadataProvidersOrderData()
```

**Review by Bond-009 — CHANGES_REQUESTED** [review]
This should fix the failing tests

**Comment by 1337joe** [comment]
> This should fix the failing tests

Thanks, I somehow missed that tests were failing after the merge from master.

**Review by Bond-009 — APPROVED** [review]

**Comment by IDisposable** [comment]
Just FYI, if building from current sources and running with external plug-ins that were compiled against the OLD `IBaseItemManager` interface, the plug-ins will throw `MissingMethodExceptions` and fail. 

I am not sure that's supposed to be a supported scenario, but adding the original methods back to the interface and implementing shim methods in the `BaseItemManager` class to "do the right thing" seems to make the plug-ins continue to function.

Just added back the old methods in the interface , and then added these forwarders.
```csharp
/// <inheritdoc />
public bool IsMetadataFetcherEnabled(BaseItem baseItem, LibraryOptions libraryOptions, string name)
{
    // just upcall into the by-TypeOptions version using the logic the old path would have used
    TypeOptions libraryTypeOptions = libraryOptions?.GetTypeOptions(baseItem.GetType().Name);
    return this.IsMetadataFetcherEnabled(baseItem, libraryTypeOptions, name);
}

/// <inheritdoc />
public bool IsImageFetcherEnabled(BaseItem baseItem, LibraryOptions libraryOptions, string name)
{
    // just upcall into the by-TypeOptions version using the logic the old path would have used
    TypeOptions libraryTypeOptions = libraryOptions?.GetTypeOptions(baseItem.GetType().Name);
    return this.IsImageFetcherEnabled(baseItem, libraryTypeOptions, name);
}
```

**Comment by IDisposable** [comment]
PR #9566 submitted to restore the old methods (as shims to the new ones)

**Pull request merged**